### DISCUSSION TITLE
Mirth Connect get stuck in reading mode

### DISCUSSION THREAD

**Discussion by alazarbaharu** [nextgenhealthcare/connect#4837] [Q&A]
Hello Everyone,

My name is Alazar I am from Ethiopia HDSS site Data Manager. We are currently using OPENHDS to register demographic data. The data collection is done by ODK. Then using mirth connect we migrate data from ODK MySQL db to OPENHDS. The version of mirth used is 3.4.2 because the migration channels are written in this version and OPENHDS only works with java 1.7.

The problem is when i deploy those channels and start the data migration Mirth connect get stuck in reading mode. I need your help on this please.

Thanks,

**Comment by jonbartels** [comment]
Can you post screenshots or logs of what happens when MC is stuck? 

Is it stuck reading on a source connector or on a destination connector?

I do not remember if the "Connection Log" panel existed in 3.4.2, it looks like this at the bottom of the message view list. It looks like this: 
<img width="1712" alt="image" src="https://user-images.githubusercontent.com/7669292/140939384-c5d6bfd9-3b2b-4f51-82fa-df51e0ea2cac.png">

If you have this connection log, post it also.

**Reply by alazarbaharu** [reply]
Dear Jon,
Here is the screenshot of Mirth when it stuck in reading mode and sometimes saying idle for a days or long and all of the sudden it will start working and again it will get in the same reading and idle mode.

![mirth log file](https://user-images.githubusercontent.com/9800211/140947232-fea47097-c183-4b32-96a9-5e7fe8cc263a.png)
![mirth reading mode](https://user-images.githubusercontent.com/9800211/140947234-f171c762-8348-4e18-b73a-c07d304146ed.PNG)

**Reply by jonbartels** [reply]
That view is helpful. I see that your "Update Events" and "Update Household" are Javascript Readers. What code do those run when the channel is called?

If a JS reader is stuck in the "reading" state it is because the code in the JS Reader itself is running or slow or stuck. This is the code in the JS Reader panel, source transformers run AFTER that initial code block runs.

You could add debug calls to log out the start time and finish time of your JS Reader code to see if it is slow.

**Reply by alazarbaharu** [reply]
The code for the "Update Events" in the source tab is 

var forms = new java.util.ArrayList();
var visitIdList = new java.util.ArrayList();

var dbConn = DatabaseConnectionFactory.createDatabaseConnection('com.mysql.jdbc.Driver', globalMap.get("ODK_DATABASE_URL"), globalMap.get("ODK_DATABASE_USER"), globalMap.get("ODK_DATABASE_PASSWORD"));

var visitResult = dbConn.executeCachedQuery("SELECT DISTINCT VISIT_DATE, OPENHDS_VISIT_ID, OPENHDS_LOCATION_ID, OPENHDS_FIELD_WORKER_ID, OPENHDS_ROUND_NUMBER, REAL_VISIT FROM " + globalMap.get("VISIT_TABLE") + " WHERE processed_by_mirth=0 OR processed_by_mirth IS NULL order by OPENHDS_VISIT_ID ASC");

while(visitResult.next()) {
  var visitXml = resultSetToXml(visitResult, "<visit event='visit' table-name='" + globalMap.get("VISIT_TABLE") + "'/>");
  // does this visit depend on a newly created location ?
  // a visit will not be processed correctly if the location it refers to does not yet exist
  var location = dbConn.executeCachedQuery("SELECT _URI, OPENHDS_HIERARCHY_ID, OPENHDS_LOCATION_ID, OPENHDS_FIELD_WORKER_ID, LOCATION_TYPE, LOCATION_NAME, GEOPOINT_ACC ACCURACY, GEOPOINT_ALT ALTITUDE, GEOPOINT_LNG LONGITUDE, GEOPOINT_LAT LATITUDE FROM " + globalMap.get("LOCATION_TABLE") + " WHERE OPENHDS_LOCATION_ID = '" + visitXml['openhds_location_id'].toString() + "' AND processed_by_mirth =0");
  
  if (location.next()) {
    var loc = resultSetToXml(location, "<location event='location' table-name='" + globalMap.get("LOCATION_TABLE") + "'/>");
    forms.add(loc.toString());
  }

  forms.add(visitXml.toString());
  
  var visitId = visitXml["openhds_visit_id"].toString();

  if(!visitIdList.contains(visitId)){
	  forms.addAll(addOutMigrations(visitId, dbConn));
	  forms.addAll(addOutMigrationsIM(visitId, dbConn));
	  forms.addAll(addInternalInmigrations(visitId, dbConn));
	  forms.addAll(addExternalInmigrations(visitId, dbConn));
	  forms.addAll(addPregnancyOutcomes(visitId, dbConn));
	  forms.addAll(addPregnancyObs(visitId, dbConn));
	  forms.addAll(addDeaths(visitId, dbConn));
	  forms.addAll(addHoHDeaths(visitId, dbConn));
	  forms.addAll(changeHoH(visitId, dbConn));
	  visitIdList.add(visitId);
  }
}


dbConn.close();
// You may access this result below with $('column_name')
return forms;

So where can i put your suggestion or the code you asked me is in different location?

**Reply by jonbartels** [reply]
You should add a logger before and after your calls to `executeCachedQuery` . Those logger statements will show you if your SQL queries are slow. A slow SQL query would cause your JS Reader to show a "Reading" status.

You also should consider:

- Refactoring your SQL queries - you should be able to use a JOIN to get the data instead of running a query in a loop
- Adding a LIMIT to your queries so you only get 50-100 rows at a time then process those in smaller sets. Your channel might be slow if you read a lot of rows to process at once

**Reply by tonygermano** [reply]
Using a JOIN and limiting the query were both suggestions that I was going to make as well.

**Reply by pacmano1** [reply]
A reminder to please format your code when you post it, see https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by alazarbaharu** [reply]
Dear Jon,
Could you please help me put the Logger on the code or give me some example to see it in action. Also can you send me a good tutorial link that can teach how to use mirth, create channels, filters...

I am asking you this because i don't have any experience with Mirth connect rather than operating on it.

The Channel we are using is developed by a team and they are no longer supporting us. Most of the HDSS sites in Ethiopia uses Mirth but we don't have any expert.

Thanks,

**Reply by pacmano1** [reply]
@alazarbaharu Please see [https://github.com/nextgenhealthcare/connect/discussions/4659] for a discussion on learning mirth.   There isn't really a concise place to learn mirth perhaps other than formal training from nextgen.

@jonbartels may chime in as well with other references.    

In my opinion it's somewhat of an obtuse tool when you start using it.   Having in a senior experienced developer on staff or as a consultant goes a very long way to smooth the learning curve.    You may find the mirth forums especially useful given its history of information.

**Reply by jonbartels** [reply]
A simple statement like `logger.error("Starting visit query...")` and `logger.error("Completed visit query")` then look at the MC logs in `mirth.log` or the Client log view and that will have the timestamp and the log message. Do the math and see how long the queries take to execute. 

You should also look at your database logs for OPENHDS. Most databases will log slow running queries.

https://forums.mirthproject.io/ is a good reference - Search for "logging" or "slow SQL" and there should be a lot of threads on how to tune it.

**Reply by alazarbaharu** [reply]
@pacmano1 Thanks for the explanation. Last time we were looking for such a consultancy firm, but we didn't take it seriously and now its affecting us a lot.  How can we get such a support? or a kind of expert training to come to Ethiopia and give a training? 

I think most of the sites agree with my suggestion so please suggest me some consultancy firms.

Thanks,

**Reply by jonbartels** [reply]
Another good resource might be the OpenHDS project itself - https://github.com/OpenHDS/Mirth-Channels and https://github.com/OpenHDS/openhds-server .

That team might not help with MC, but they might have advice for how to efficiently query OpenHDS for the data you need.

**Reply by pacmano1** [reply]
My company, https://diridium.com/, is a USA based consultancy.   Regardless of that mention, the best of the best consultants can pricey which I think that is mostly a reflection of the amount of work in the healthcare integration space today.    Of course "pricey" can be relative to a lot of things!     I am sure some other consultancies might chime in here.

It's likely a topic to take off line.

**Reply by alazarbaharu** [reply]
@jonbartels  and @pacmano1 I would like to say thanks for your kind support. Will dig in to it and will comeback to you.

Once again thanks for the support it means a lot to me.
